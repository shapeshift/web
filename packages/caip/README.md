# @shapeshiftoss/caip

This package is ShapeShift's partial implementation of [CAIPs](https://github.com/ChainAgnostic/CAIPs) - Chain Agnostic Improvement Protocols.
It is not exhaustive and is currently only used internally.

## ChainId (CAIP2) - Blockchain ID Specification

https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-2.md

Usage

### `toChainId` | `toCAIP2`

```ts
const chainNamespace = CHAIN_NAMESPACE.Evm
const chainReference = CHAIN_REFERENCE.EthereumMainnet
const result = toChainId({ chainNamespace, chainReference })
expect(result).toEqual('eip155:1')
```

### `fromChainId` | `fromCAIP2`

```ts
const ethereumChainId = 'eip155:1'
const { chainNamespace, chainReference } = fromChainId(ethereumChainId)
expect(chainNamespace).toEqual(CHAIN_NAMESPACE.Evm)
expect(chainReference).toEqual(CHAIN_REFERENCE.EthereumMainnet)
```

## AccountId (CAIP10) - Blockchain ID Specification

https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-10.md

Usage

### `toAccountId` | `toCAIP10`

```ts
const chainNamespace = CHAIN_NAMESPACE.Evm
const chainReference = CHAIN_REFERENCE.EthereumMainnet
const chainId = toChainId({ chainNamespace, chainReference })
const account = '0xa44c286ba83bb771cd0107b2c1df678435bd1535'
const expectedAccountId = `${chainId}:${account}`
expect(toAccountId({ chainId, account })).toEqual('eip155:1:0xa44c286ba83bb771cd0107b2c1df678435bd1535')
expect(toAccountId({ chainNamespace, chainReference, account })).toEqual('eip155:1:0xa44c286ba83bb771cd0107b2c1df678435bd1535')
```

### `fromAccountId` | `fromCAIP10`

```ts
const accountId = 'eip155:1:0xa44c286ba83bb771cd0107b2c1df678435bd1535'
const { account, chainId, chainNamespace, chainReference } = fromAccountId(accountId)
expect(account).toEqual(0xa44c286ba83bb771cd0107b2c1df678435bd1535)
expect(chainNamespace).toEqual('eip155')
expect(chainReference).toEqual('1')
expect(chainId).toEqual('eip155:1')
```

## AssetId (CAIP19) - Asset Type and Asset ID Specification

https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-19.md

Usage

### `toAssetId` | `toCAIP19`

Ether

```ts
const chainNamespace = CHAIN_NAMESPACE.Evm
const chainReference = CHAIN_REFERENCE.EthereumMainnet
const chainId = toChainId({ chainNamespace, chainReference })
const assetNamespace = 'slip44'
const assetReference = ASSET_REFERENCE.Ethereum
expect(toAssetId({ chainNamespace, chainReference, assetNamespace, assetReference })).toEqual('eip155:1/slip44:60')
expect(toAssetId({ chainId, assetNamespace, assetReference })).toEqual('eip155:1/slip44:60')
```

ERC20 token

```ts
const chainNamespace = CHAIN_NAMESPACE.Evm
const chainReference = CHAIN_REFERENCE.EthereumMainnet
const chainId = toChainId({ chainNamespace, chainReference })
const assetNamespace = 'erc20'
const assetReference = '0xc770eefad204b5180df6a14ee197d99d808ee52d'
expect(toAssetId({ chainNamespace, chainReference, assetNamespace, assetReference })).toEqual('eip155:1/erc20:0xc770eefad204b5180df6a14ee197d99d808ee52d')
expect(toAssetId({ chainId, assetNamespace, assetReference })).toEqual('eip155:1/erc20:0xc770eefad204b5180df6a14ee197d99d808ee52d')
```

### `fromAssetId` | `fromCAIP19`

Ether

```ts
const assetId = 'eip155:1/slip44:60'
const { chainId, chainReference, chainNamespace, assetNamespace, assetReference } = fromAssetId(AssetId)
expect(chainNamespace).toEqual('eip155')
expect(chainReference).toEqual('1')
expect(chainId).toEqual('eip155:1')
expect(assetNamespace).toEqual('slip44')
expect(assetReference).toEqual('60')
```

ERC20 token

```ts
const assetId = 'eip155:1/erc20:0xc770eefad204b5180df6a14ee197d99d808ee52d'
const { chainId, chainReference, chainNamespace, assetNamespace, assetReference } = fromAssetId(AssetId)
expect(chainNamespace).toEqual('eip155')
expect(chainReference).toEqual('1')
expect(chainId).toEqual('eip155:1')
expect(assetNamespace).toEqual('erc20')
expect(assetReference).toEqual('0xc770eefad204b5180df6a14ee197d99d808ee52d')
```

## Adapters

Adapters map asset IDs and map them to vendor-specific (e.g. CoinGecko) IDs.

This allows us to use asset IDs internally and keep any vendor IDs at the boundary.

### Generate

To generate new static adapter data, run the following

```zsh
cd packages/caip
yarn generate
```

and commit the generated `adapter.json` files.

### Usage

```ts
console.log(coingeckoToAssetIds('shapeshift-fox-token'))

[eip155:1/erc20:0xc770eefad204b5180df6a14ee197d99d808ee52d]

console.log(assetIdToCoingecko('bip122:000000000019d6689c085ae165831e93/slip44:0'))

bitcoin
```
