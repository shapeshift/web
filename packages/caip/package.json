{
  "name": "@shapeshiftoss/caip",
  "version": "8.15.0",
  "description": "CAIP Implementation",
  "homepage": "",
  "license": "MIT",
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "files": [
    "dist"
  ],
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/shapeshift/lib"
  },
  "scripts": {
    "build": "yarn clean && yarn compile",
    "clean": "rm -rf dist && rm -rf tsconfig.build.tsbuildinfo",
    "compile": "tsc -p tsconfig.build.json",
    "dev": "tsc --watch",
    "generate-coingecko": "ts-node -P tsconfig.build.json src/adapters/coingecko/generate.ts",
    "generate-coincap": "ts-node -P tsconfig.build.json src/adapters/coincap/generate.ts",
    "generate-yearn": "ts-node -P tsconfig.build.json src/adapters/yearn/generate.ts",
    "generate-osmosis": "ts-node -P tsconfig.build.json src/adapters/osmosis/generate.ts",
    "generate": "yarn generate-coingecko && yarn generate-coincap && yarn generate-yearn && yarn generate-osmosis",
    "prepare": "yarn build",
    "test": "jest test",
    "type-check": "tsc --project ./tsconfig.build.json --noEmit"
  },
  "peerDependencies": {
    "@shapeshiftoss/types": "^8.1.0"
  },
  "devDependencies": {
    "@shapeshiftoss/types": "^8.1.0",
    "@yfi/sdk": "^1.2.0",
    "axios": "^0.26.1",
    "ts-node": "^10.7.0"
  }
}
