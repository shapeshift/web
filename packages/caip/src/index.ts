export * as adapters from './adapters'
export * from './chainId/chainId'
export * from './accountId/accountId'
export * from './assetId/assetId'
export * from './utils'
export * from './constants'
export * from './typeGuards'
