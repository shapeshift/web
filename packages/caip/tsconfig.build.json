{
  "extends": "../../tsconfig.build.json",
  "compilerOptions": {
    "composite": true,
    "outDir": "./dist",
    "rootDir": "src",
    "allowJs": true,
  },
  "include": ["src/**/*", "src/**/*.json"],
  "exclude": ["node_modules", "**/*.test.ts", "**/*.spec.ts"],
  "references": [
    { "path": "../types/tsconfig.build.json" },
  ],
}
