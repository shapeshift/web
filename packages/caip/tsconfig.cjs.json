{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "module": "CommonJS",
    "moduleResolution": "node",
    "rootDir": "src",
    "outDir": "dist/cjs",
    "tsBuildInfoFile": "dist/tsconfig.cjs.tsbuildinfo",
    "noEmit": false
  },
  "include": ["src/**/*", "src/**/*.json"],
  "exclude": ["dist"]
}
