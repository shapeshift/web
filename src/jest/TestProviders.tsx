import { translations } from 'assets/translations'
import { I18n } from 'react-polyglot'

const locale: string = navigator?.language?.split('-')[0] ?? 'en'
const messages = translations[locale]

export const TestProviders: React.FC = ({ children }) => (
  <I18n locale={locale} messages={messages}>
    {children}
  </I18n>
)
