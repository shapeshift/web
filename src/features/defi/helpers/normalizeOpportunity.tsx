import type { AssetId, ChainId } from '@shapeshiftoss/caip'
import { fromAssetId } from '@shapeshiftoss/caip'
import { USDC_PRECISION } from 'constants/constants'
import { useTranslate } from 'react-polyglot'
import { useSelector } from 'react-redux'
import { bnOrZero } from 'lib/bignumber/bignumber'
import type { MergedActiveStakingOpportunity } from 'pages/Defi/hooks/useCosmosSdkStakingBalances'
import type { MergedSerializableOpportunity } from 'pages/Defi/hooks/useVaultBalances'
import { useVaultBalances } from 'pages/Defi/hooks/useVaultBalances'
import type { MergedFoxyOpportunity } from 'state/apis/foxy/foxyApi'
import { selectAssetIds } from 'state/slices/selectors'

import { DefiType } from '../contexts/DefiManagerProvider/DefiCommon'
import { chainIdToLabel } from './utils'

export type EarnOpportunityType = {
  type?: string
  provider: string
  version?: string
  contractAddress: string
  rewardAddress: string
  apy?: number | string
  tvl: string
  underlyingAssetId?: AssetId
  assetId: AssetId
  fiatAmount: string
  cryptoAmount: string
  expired?: boolean
  chainId: ChainId
  moniker?: string
  showAssetSymbol?: boolean
  isLoaded: boolean
  icons?: string[]
  // overrides any name down the road
  opportunityName?: string
  highestBalanceAccountAddress?: string // FOX/ETH specific, let's change it to accountId across the line if we need it for other opportunities
}

const useTransformVault = (vaults: MergedSerializableOpportunity[]): EarnOpportunityType[] => {
  const assetIds = useSelector(selectAssetIds)

  const { vaults: vaultsWithBalances } = useVaultBalances()
  return vaults.reduce<EarnOpportunityType[]>((acc, vault) => {
    const chainId = fromAssetId(vault.feeAsset.assetId).chainId
    let fiatAmount = '0'
    let cryptoAmount = '0'
    if (vaultsWithBalances[vault.id]) {
      const balances = vaultsWithBalances[vault.id]
      cryptoAmount = balances.cryptoAmount
      fiatAmount = balances.fiatAmount
    }
    const assetId = vault.underlyingAsset.assetId
    const data = {
      type: DefiType.Vault,
      provider: vault.provider,
      version: vault.version,
      contractAddress: vault.id,
      rewardAddress: vault.id,
      tvl: bnOrZero(vault.tvl.balanceUsdc).div(`1e+${USDC_PRECISION}`).toString(),
      apy: vault.apy.toString(),
      expired: vault?.expired || false,
      chainId,
      assetId,
      fiatAmount,
      cryptoAmount,
      // DeFi foxy and yearn vaults are already loaded by the time they are transformed
      isLoaded: true,
    }
    // show vaults that are expired but have a balance
    // show vaults that don't have an APY but have a balance
    // don't show vaults that don't have a balance and don't have an APY
    // don't show new vaults that have an APY over 20,000% APY
    if (assetIds.includes(assetId)) {
      if (
        vault?.expired ||
        bnOrZero(vault?.apy).isEqualTo(0) ||
        bnOrZero(vault.tvl.balanceUsdc).isEqualTo(0) ||
        (bnOrZero(vault?.apy).gt(200) && (vault?.isNew || false))
      ) {
        if (bnOrZero(cryptoAmount).gt(0)) {
          acc.push(data)
        }
      } else {
        acc.push(data)
      }
    }
    return acc
  }, [])
}

const transformFoxy = (foxies: MergedFoxyOpportunity[]): EarnOpportunityType[] =>
  foxies.map(foxy => {
    const {
      provider,
      contractAddress,
      stakingToken: tokenAddress,
      rewardToken: rewardAddress,
      tvl,
      apy,
      expired,
      chainId,
      tokenAssetId: assetId,
      fiatAmount,
      cryptoAmount,
    } = foxy
    return {
      type: DefiType.TokenStaking,
      provider,
      contractAddress,
      tokenAddress,
      rewardAddress,
      tvl: bnOrZero(tvl).toString(),
      apy,
      expired,
      chainId,
      assetId,
      fiatAmount,
      cryptoAmount,
      // DeFi foxy and yearn vaults are already loaded by the time they are transformed
      isLoaded: true,
    }
  })

const useTransformCosmosStaking = (
  cosmosStakingOpportunities: MergedActiveStakingOpportunity[],
): EarnOpportunityType[] => {
  const translate = useTranslate()
  return cosmosStakingOpportunities
    .map(staking => {
      return {
        type: DefiType.TokenStaking,
        provider: chainIdToLabel(staking.chainId),
        contractAddress: staking.address,
        rewardAddress: '',
        tvl: staking.tvl,
        apy: staking.apr,
        chainId: staking.chainId,
        assetId: staking.assetId,
        fiatAmount: staking.fiatAmount ?? '',
        cryptoAmount: staking.cryptoAmount ?? '',
        moniker: staking.moniker,
        version:
          !bnOrZero(staking.cryptoAmount).isZero() &&
          translate('defi.validatorMoniker', { moniker: staking.moniker }),
        showAssetSymbol: bnOrZero(staking.cryptoAmount).isZero(),
        isLoaded: Boolean(staking.isLoaded),
      }
    })
    .sort((opportunityA, opportunityB) => {
      return bnOrZero(opportunityA.cryptoAmount).gt(bnOrZero(opportunityB.cryptoAmount)) ? -1 : 1
    })
}

type NormalizeOpportunitiesProps = {
  vaultArray: MergedSerializableOpportunity[]
  foxyArray: MergedFoxyOpportunity[]
  cosmosSdkStakingOpportunities: MergedActiveStakingOpportunity[]
  foxEthLpOpportunity?: EarnOpportunityType
  stakingOpportunities?: EarnOpportunityType[]
}

export const useNormalizeOpportunities = ({
  vaultArray,
  foxyArray,
  cosmosSdkStakingOpportunities = [],
  foxEthLpOpportunity,
  stakingOpportunities,
}: NormalizeOpportunitiesProps): EarnOpportunityType[] => {
  return [
    ...transformFoxy(foxyArray),
    ...useTransformCosmosStaking(cosmosSdkStakingOpportunities),
    ...(stakingOpportunities ? stakingOpportunities : []),
    ...(foxEthLpOpportunity ? [foxEthLpOpportunity] : []),
    ...useTransformVault(vaultArray),
  ]
}
