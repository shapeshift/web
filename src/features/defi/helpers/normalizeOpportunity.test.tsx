import { cosmosChainId } from '@shapeshiftoss/caip'
import { renderHook } from '@testing-library/react'
import type { PropsWithChildren } from 'react'
import { TestProviders } from 'test/TestProviders'
import type { MergedActiveStakingOpportunity } from 'pages/Defi/hooks/useCosmosSdkStakingBalances'

import { useNormalizeOpportunities } from './normalizeOpportunity'

jest.mock('@shapeshiftoss/investor-yearn')
jest.mock('pages/Defi/hooks/useFoxyBalances')

const mockCosmosStakingOpportunities = [
  {
    address: 'cosmosvaloper199mlc7fr6ll5t54w7tts7f4s0cvnqgc59nmuxf',
    moniker: 'ShapeShift DAO',
    tokens: '42424242',
    commission: '0.100000000000000000',
    apr: '0.1528209855',
    totalDelegations: '42',
    rewards: '4.2',
    isLoaded: true,
    cryptoAmountBaseUnit: '1337',
    cryptoAmountPrecision: '0.001337',
    tvl: '21040543.6367982',
    fiatAmount: '4.2',
    chainId: cosmosChainId,
    assetId: 'cosmos:cosmoshub-4/slip44:118',
    tokenAddress: '118',
  },
  {
    address: 'cosmosvaloper1clpqr4nrk4khgkxj78fcwwh6dl3uw4epsluffn',
    moniker: 'Cosmostation',
    tokens: '24242424',
    commission: '0.089000000000000000',
    apr: '0.1546887975',
    totalDelegations: '242424',
    rewards: '2.5',
    isLoaded: true,
    cryptoAmountBaseUnit: '12345',
    cryptoAmountPrecision: '0.012345',
    tvl: '63799889.014332',
    fiatAmount: '0.24',
    chainId: cosmosChainId,
    assetId: 'cosmos:cosmoshub-4/slip44:118',
    tokenAddress: '118',
  },
]
function setup({
  cosmosStakingOpportunities,
}: {
  cosmosStakingOpportunities?: MergedActiveStakingOpportunity[]
} = {}) {
  const wrapper: React.FC<PropsWithChildren> = ({ children }) => (
    <TestProviders>{children}</TestProviders>
  )
  const { result } = renderHook(
    () =>
      useNormalizeOpportunities({
        cosmosSdkStakingOpportunities: cosmosStakingOpportunities ?? [],
        foxyArray: [],
      }),
    { wrapper },
  )
  return { result }
}

describe('useNormalizeOpportunities', () => {
  it('returns empty arrays when provided with empty arrays', () => {
    const { result } = setup()
    expect(result.current).toEqual([])
  })

  it('returns transformed array of active opportunities sorted by cryptoAmount when there are active staking opportunities', () => {
    const { result } = setup({
      cosmosStakingOpportunities: mockCosmosStakingOpportunities,
    })
    expect(result.current).toMatchSnapshot()
  })

  it('returns transformed array of staking opportunities when there is no active staking opportunity', () => {
    const { result } = setup({ cosmosStakingOpportunities: [] })
    expect(result.current).toMatchSnapshot()
  })
})
