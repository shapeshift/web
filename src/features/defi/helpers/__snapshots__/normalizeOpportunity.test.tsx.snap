// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useNormalizeOpportunities returns transformed array of active opportunities sorted by cryptoAmount when there are active staking opportunities 1`] = `
Array [
  Object {
    "apy": "0.1546887975",
    "assetId": "cosmos:cosmoshub-4/slip44:118",
    "chainId": "cosmos:cosmoshub-4",
    "contractAddress": "cosmosvaloper1clpqr4nrk4khgkxj78fcwwh6dl3uw4epsluffn",
    "cryptoAmountBaseUnit": "12345",
    "cryptoAmountPrecision": "0.012345",
    "fiatAmount": "0.24",
    "isLoaded": true,
    "moniker": "Cosmostation",
    "provider": "Cosmos",
    "rewardAddress": "",
    "showAssetSymbol": false,
    "tvl": "63799889.014332",
    "type": "token_staking",
    "version": "Cosmostation Validator",
  },
  Object {
    "apy": "0.1528209855",
    "assetId": "cosmos:cosmoshub-4/slip44:118",
    "chainId": "cosmos:cosmoshub-4",
    "contractAddress": "cosmosvaloper199mlc7fr6ll5t54w7tts7f4s0cvnqgc59nmuxf",
    "cryptoAmountBaseUnit": "1337",
    "cryptoAmountPrecision": "0.001337",
    "fiatAmount": "4.2",
    "isLoaded": true,
    "moniker": "ShapeShift DAO",
    "provider": "Cosmos",
    "rewardAddress": "",
    "showAssetSymbol": false,
    "tvl": "21040543.6367982",
    "type": "token_staking",
    "version": "ShapeShift DAO Validator",
  },
]
`;

exports[`useNormalizeOpportunities returns transformed array of staking opportunities when there is no active staking opportunity 1`] = `Array []`;
